### PR TITLE
Fix Hadoop conf files loading/processing in the Compaction tool

### PR DESCRIPTION
**Pull request by lbendig**
This patch addresses the following issues:

1.
Fix conf files loading:
The Configuration object is populated from the Hadoop conf xmls defined in the ```hadoop.configfile.*``` properties. However, ```HdfsIO#addHadoopConfigPropertiesToConf``` calls
 ```Configuration#addResource(String)```. This method only examines the classpath for the given file name. Thus if the file is not on the classpath, the Configuration object will remain empty

2.
Ensure the correct order of the HDFS_URI resolution:
If ```fs.defaultFS``` is set to the default value before the conf files are loaded, then it won't get overridden by either the value in ```core-site.xml``` or in ```hdfs.uri```. The resolved ```hdfs.uri``` should not only be set to the Configuration object, but to the ```CompactionRunner.properties```, because ```HdfsIO#getHdfsUri``` use that object for the resolution.

3.
Load Hive JDBC driver with the default hiveserver.version value if it is not set:
The docs states that the default ```hiveserver.version``` is 2. But if it is not set by the user, the ```HiveJdbcConnector#newConnectorWithProps``` will not load the driver.

4.
Fix the splitting of comma separated column names in snapshot.pkey and delta.i.pkey:
```CompactionRunner#buildAvroExternalTable``` passes ```jobProperties.getProperty(tableType + PKEY))``` as parameter to ```HiveTable#withPrimaryKeys(String firstKeyAttr, String... remainingKeyAttr)``` . Even if the parameter contains commas, it won't be split automaticaly 
between the first and variable length paramteres; ```firstKeyAttr``` will store the whole string.

5.
If multiple primary keys are defined in the ```task.conf```, the LEFT JOIN and the WHERE clause of the INSERT OVERWRITE statement will be broken, e.g:
INSERT OVERWRITE TABLE not_updated
SELECT merged_delta.* FROM merged_delta LEFT OUTER JOIN delta2 ON merged_delta.my_id1,my_id2 = delta2.my_id1,my_id2
WHERE delta2.my_id1,my_id2 IS NULL

### CODE REVIEW

**Comment by sahilTakiar** [comment]
LGTM. @zliu41 any comments?

File: gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java [hunk 1]
```diff
@@ -84,11 +86,15 @@ protected static Configuration getConfiguration() throws FileNotFoundException {
...
 
   private static void addResourceToConf(Configuration conf) {
-    conf.setStrings(getHdfsUriHadoopPropertyName(), HDFS_URI_DEFAULT);
     addHadoopConfigPropertiesToConf(conf);
+    String hdfsUriKey = getHdfsUriHadoopPropertyName();
     if (CompactionRunner.properties.containsKey(HDFS_URI)) {
-      conf.setStrings(getHdfsUriHadoopPropertyName(), CompactionRunner.properties.getProperty(HDFS_URI));
+      conf.set(hdfsUriKey, CompactionRunner.properties.getProperty(HDFS_URI));
     }
+    if (Strings.isNullOrEmpty(conf.get(hdfsUriKey))) {
```
Comment (zliu41) [inline gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java]: hdfsUriKey can't be null or empty here (and if it is, then the next line doesn't make sense).
I think you can replace the `if` block in line 91-93 with `conf.set(hdfsUriKey, CompactionRunner.properties.getProperty(HDFS_URI, HDFS_URI_DEFAULT));`

**Comment by zliu41** [comment]
LGTM barring the above comment.

File: gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java (hunk 1 above)
Comment (lbendig) [inline gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java]: Thanks for your suggestions. I assume that the order of setting ```fs.defaultFS``` is that first we're loading it from the hadoop conf XMLs, then if we see that ```hdfs.uri``` has been defined in the compaction job conf, we use that, and at last we fall back to the default.
If I replace lines 91-93, won't we override the value taken from the hadoop conf with the default one each time if ```hdfs.uri``` is not set?

**Comment by lbendig** [comment]
@zliu41 What do you think?

File: gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java (hunk 1 above)
Comment (zliu41) [inline gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java]: You are right, I read it wrong.

Comment (lbendig) [inline gobblin-compaction/src/main/java/gobblin/compaction/HdfsIO.java]: Thanks for the review!

**Pull request merged**